source/registry.d

```diff
@@ -0,0 +1,679 @@
+ *     higher than T.max; to complement these default type constraints you cannot
+ *     pass a value equal to 0 as this would mean no space for ids or equal to
+ *     T.sizeof * 8 as this would mean no space for batches. \
+ * \
+ * There are some default alises available for fast access.
+ *
+ * | alias-name | id-(bits) | batch-(bits) | max-entities  | batch-reset   |
+ * | :--------- | :-------: | :----------: | :-----------: | :-----------: |
+ * | Registry   | 20        | 12           | 1_048_574     | 4_096         |
+ * | Registry8  | 4         | 4            | 14            | 16            |
+ * | Registry16 | 8         | 8            | 254           | 256           |
+ * | Registry32 | 16        | 16           | 65_534        | 65_536        |
+ * | Registry64 | 32        | 32           | 4_294_967_295 | 4_294_967_296 |
+ *
+ * Registry:
+ * + id: ***20 bits*** (max at `1_048_574`)
+ * + batch: ***12 bits*** (resets at `4_096`)
+ *
+ * Registry8:
+ * + id: ***4 bits*** (max at `14`)
+ * + batch: ***4 bits*** (resets at `16`)
+ *
+ * Registry16:
+ * + id: ***8 bits*** (max at `254`)
+ * + batch: ***8 bits*** (resets at `256`)
+ *
+ * Registry32:
+ * + id: ***16 bits*** (max at `65_534`)
+ * + batch: ***16 bits*** (resets at `65_536`)
+ *
+ * Registry64:
+ * + id: ***32 bits*** (max at `4_294_967_295`)
+ * + batch: ***32 bits*** (resets at `4_294_967_296`)
+ *
+ * Params:
+ *     T = entity type.
+ *     amount = number of bits reserved for the id.
+ *
+ * Code_Generation:
+ * + `entityShift` = the shift value needed to access the batch.
+ * + `entityMask` = the mask to extract the entity id.
+ * + `batchMask` = the mask to extract the entity batch.
+ * + `entityNull` = reserved value to for representing an empty queue as well as
+ *     the maximum of entities.
+ *
+ * See_Also: $(LREF isEntityType), $(LREF isValidBitIdAmount)
+ */
+private mixin template genEntityBitMasks(T, T amount)
+	if(isEntityType!T && isValidBitIdAmount!(T, amount))
+{
+	public enum T entityShift = amount;
+	public enum T entityMask = (1UL << amount) - 1;
+	public enum T batchMask = (1UL << (T.sizeof * 8 - amount)) - 1;
+	public enum T entityNull = entityMask;
+}
+
+
+/**
+ * Simple exception for entity errors.
+ */
+class EntityException : Exception
+{
+	mixin basicExceptionCtors;
+}
+
+
+@nogc @safe pure
+class BasicRegistry(T, T idBitQuantity)
+	if(isEntityType!T)
+{
+	import std.container : Array;
+	import std.conv : to;
+	import std.exception : enforce;
+
+	mixin genEntityBitMasks!(T, idBitQuantity);
+
+public:
+	this() {}
+
+
+	/**
+	 * Create a new entity either by spawning a new id or reusing one in the
+	 *     queue.
+	 *
+	 * Returns: a new entity reference with a reused id if the queue isn't
+	 *     empty or with a new spawned id otherwise.
+	 *
+	 * See_Also: $(LREF spawn), $(LREF revive), $(LREF murder)
+	 */
+	@safe pure
+	@property const(T) create()
+	{
+		return queue == entityNull ? spawn : revive;
```

Comment:
I think it doesn't hurt to add `()` so it's clear this is a function call, but it's just a minor nitpick.

Reply:
The functions are declared as @property. Create returns an entity, the user doesn't need to know how it works, only that it returns something or is something. For this reason I picked this route, declaring these functions as @property. This user can then choose what it's best for him. For me it's easier to use them as a property.

Reply:
Ah, sorry, I mean the calls to `spawn` and `revive` inside the function :stuck_out_tongue:

Reply:
done